### PR TITLE
[clang-cpp-frontend] Add CXXThisExpr and CXXConstructor

### PR DESCRIPTION
**Pull request by kunjsong01**
Extend clang-cpp-frontend to pass Constructor1-1

### CODE REVIEW

**Comment by kunjsong01** [comment]
### Goals:
Add `this' pointer as the first argument for constructor and generate the symbol table as the evience to show:
1. we have successfully generated the symbol for constructor function
2. the symbol has a parameter `this'
3. the constructor function is properly called in the main function `symbol.value'

### What's covered in previous patches:
(see commit message) 

### Current progress: 
```
Converting
TranslationUnit has a total of 12 declarations
converting class symbol
RecordDecl has 1 fields
RecordDecl has 1 methods
Pointer `this' for method t1 was not added to scope ---> called when parsing `clang::Stmt::CXXThisExprClass'
```

### Preliminary Analysis: 
When parsing a constructor method, `get_function_this_pointer_param` shall be called. 
`this_map` shall be populated by `get_function_this_pointer_param`, currently missing in `clang-cpp-frontend`. 

### TODOs:
1.  ~~Split `get_function_params` in `clang_c_frontend`. Make sure we are not breaking regressions for C verification.~~ (done in commit a49695a) 
2. ~~Add `get_function_this_pointer_param` to update `this_map`~~ (done in commit c8efa3d8835, note: id==prefix+name in the src_v2)
3. ~~Add `get_function_params` in `clang_cpp_frontend` to call `get_function_this_pointer_param`~~ (done in commit d8989e8)

**Comment by kunjsong01** [comment]
### Goal update:

1. ~~we have successfully generated the symbol for constructor function~~ (done, as shown above)
2. ~~the symbol has a parameter `this'~~ (done, as shown above)
3. ~~the constructor function is properly called in the main function `symbol.value'~~ (done, see [comment](https://github.com/esbmc/esbmc/pull/712#issuecomment-1138952645))

```
728 Symbol......: c:@S@t1@F@t1#
729 Module......: main
730 Base name...: t1
731 Mode........: C (0)
732 Type........: void ( t1 *)
733 Value.......:
734 {
735 t1->i = 1;
736
737 }
738 Flags.......: lvalue
739 Location....: file main.cpp line 7
```

### TODOs:
1. ~~Populate side_effect of the constructor symbol~~ (done in commit b0cc63eac)
2. ~~To complete complete Goal-3, add CXXConstructorExpr~~ (done in commit b0cc63eac)

**Comment by kunjsong01** [comment]
### Constructor callsite in main symbol:
```
718 Symbol......: c:@F@main#
719 Module......: main
720 Base name...: main
721 Mode........: C (0)
722 Type........: signed int ()
723 Value.......:
724 {
725 t1 instance1;
726 t1(&instance1);
727
728
729 }
```

### Constructor callsite in main GOTO:
```
 68 main (c:@F@main#):
 69         // 22 no location
 70         t1 instance1;
 71         // 23 no location
 72         FUNCTION_CALL:  t1(&instance1)
 73         // 24 file main.cpp line 18 function main
 74         dead instance1;
 75         // 25 file main.cpp line 18 function main
 76         RETURN: NONDET(signed int)
 77         // 26 file main.cpp line 18 function main
 78         END_FUNCTION // main
 79 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Test case Constructor1-1 is now passing. 

TODOs:
- ~~code clean-up~~ (done in commit 79caab6e14)

**Comment by kunjsong01** [comment]
Disabled a bunch of test cases in regression: 
```
	1711 - regression/esbmc-cpp/cbmc/Templates14 (Failed)
	1713 - regression/esbmc-cpp/cbmc/Templates16 (Failed)
	1714 - regression/esbmc-cpp/cbmc/Templates17 (Failed)
	1715 - regression/esbmc-cpp/cbmc/Templates18 (Failed)
	1717 - regression/esbmc-cpp/cbmc/Templates2 (Failed)
	1733 - regression/esbmc-cpp/cbmc/Templates34 (Failed)
	1735 - regression/esbmc-cpp/cbmc/Templates36 (Failed)
	1736 - regression/esbmc-cpp/cbmc/Templates37 (Failed)
	1738 - regression/esbmc-cpp/cbmc/Templates39 (Failed)
	1742 - regression/esbmc-cpp/cbmc/Templates7 (Failed)
```

They were enabled as a result of commit b179fb5240ac, in which the CXXConstructExpr was handled in a hacky way.

File: src/clang-cpp-frontend/clang_cpp_convert.h [hunk 1]
```diff
@@ -16,14 +16,35 @@ class clang_cpp_convertert : public clang_c_convertert
   virtual ~clang_cpp_convertert() = default;
 
 protected:
+  typedef std::unordered_map<std::size_t, std::pair<std::string, typet>>
+    this_mapt;
+  this_mapt this_map;
+
   bool get_decl(const clang::Decl &decl, exprt &new_expr) override;
 
+  bool get_decl_ref(const clang::Decl &decl, exprt &new_expr);
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.h:30]: Shall we document this method?

File: src/clang-cpp-frontend/clang_cpp_convert.h [hunk 2]
```diff
@@ -16,14 +16,35 @@ class clang_cpp_convertert : public clang_c_convertert
...
+
   bool get_type(const clang::QualType &type, typet &new_type) override;
 
   bool get_type(const clang::Type &the_type, typet &new_type) override;
 
   bool get_function(const clang::FunctionDecl &fd, exprt &new_expr) override;
 
+  bool get_constructor_call(
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.h:41]: Shall we document this method?

File: src/clang-cpp-frontend/clang_cpp_convert.h [hunk 3]
```diff
@@ -16,14 +16,35 @@ class clang_cpp_convertert : public clang_c_convertert
...
+    const clang::CXXConstructExpr &constructor_call,
+    exprt &new_expr);
+
+  bool
+  get_function_body(const clang::FunctionDecl &fd, exprt &new_expr) override;
+
+  bool get_function_params(
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.h:52]: Shall we document these news methods?

File: src/clang-cpp-frontend/clang_cpp_convert.h [hunk 4]
```diff
@@ -16,14 +16,35 @@ class clang_cpp_convertert : public clang_c_convertert
...
+    const clang::FunctionDecl &fd,
+    code_typet::argumentst &params) override;
+
+  bool get_function_this_pointer_param(
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.h:63]: Shall we document this method?

File: src/clang-cpp-frontend/clang_cpp_adjust_code.cpp [hunk 5]
```diff
@@ -127,3 +139,51 @@ void clang_cpp_adjust::adjust_for(codet &code)
...
+
+        // First, create new decl without rhs
+        code_declt object(code_decl.lhs());
+        new_block.copy_to_operands(object);
+
+        // Get rhs
+        side_effect_expr_function_callt &init =
+          to_side_effect_expr_function_call(rhs);
+
+        init.arguments().push_back(address_of_exprt(code_decl.lhs()));
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_adjust_code.cpp:174]: Add a comment here to keep the standard of the other lines?

File: src/clang-c-frontend/clang_c_convert.h [hunk 6]
```diff
@@ -91,7 +93,14 @@ class clang_c_convertert
 
   virtual bool get_function(const clang::FunctionDecl &fd, exprt &new_expr);
 
-  bool get_function_params(const clang::ParmVarDecl &pd, exprt &param);
+  virtual bool
+  get_function_body(const clang::FunctionDecl &fd, exprt &new_expr);
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.h:103]: Shall we document these new methods?

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 7]
```diff
@@ -587,7 +592,40 @@ bool clang_c_convertert::get_function(const clang::FunctionDecl &fd, exprt &)
...
+  const clang::FunctionDecl &fd,
+  code_typet::argumentst &params)
+{
+  if(!fd.parameters().size()) // return if no parameter
+    return false;
+
+  for(auto const &pdecl : fd.parameters())
+  {
+    code_typet::argumentt param;
+    if(get_function_param(*pdecl, param))
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp:613]: In which case do we return true?

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 8]
```diff
@@ -587,7 +592,40 @@ bool clang_c_convertert::get_function(const clang::FunctionDecl &fd, exprt &)
...
 }
 
+bool clang_c_convertert::get_function_body(
+  const clang::FunctionDecl &fd,
+  exprt &new_expr)
+{
+  assert(fd.hasBody());
+
+  exprt body_exprt;
+  if(get_expr(*fd.getBody(), body_exprt))
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp:596]: Which case do we return true?

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 9]
```diff
@@ -500,8 +514,8 @@ bool clang_c_convertert::get_var(const clang::VarDecl &vd, exprt &new_expr)
 
 bool clang_c_convertert::get_function(const clang::FunctionDecl &fd, exprt &)
 {
-  // Don't convert if clang thinks that the functions was implicitly converted
-  if(fd.isImplicit())
+  // Don't convert if implicit, unless it's a constructor
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp:509]: Why not? give some rationale here.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 10]
```diff
@@ -214,14 +214,25 @@ bool clang_c_convertert::get_decl(const clang::Decl &decl, exprt &new_expr)
...
+      "TranslationUnit has a total of {} top-level declarations",
+      std::distance(std::cbegin(tu.decls()), std::cend(tu.decls()))));
+    static unsigned ctr = 0;
     for(auto const &decl : tu.decls())
     {
       // This is a global declaration (variable, function, struct, etc)
       // We don't need the exprt, it will be automatically added to the
       // context
       exprt dummy_decl;
+      if(ctr == 10)
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp]: Why 10?

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 11]
```diff
@@ -214,14 +214,25 @@ bool clang_c_convertert::get_decl(const clang::Decl &decl, exprt &new_expr)
...
+      {
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp]: remove brackets...

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

File: src/clang-cpp-frontend/clang_cpp_convert.h (hunk 2 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.h:41]: Done

File: src/clang-cpp-frontend/clang_cpp_convert.h (hunk 1 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.h:30]: Done

File: src/clang-cpp-frontend/clang_cpp_convert.h (hunk 3 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.h:52]: Done

File: src/clang-cpp-frontend/clang_cpp_convert.h (hunk 4 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.h:63]: Done

File: src/clang-cpp-frontend/clang_cpp_adjust_code.cpp (hunk 5 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_adjust_code.cpp:174]: Done

File: src/clang-c-frontend/clang_c_convert.h (hunk 6 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.h:103]: Done. 
Explained why they are made virtual.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 7 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp:613]: Done.
Added a comment to explain a failed case.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 8 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp:596]: Done. 
Added a comment to explain a failed case

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 9 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp:509]: Done.
Also added a regression for implicit constructor.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 10 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp]: Removed. 
Sorry this was for debugging purposes.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 11 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp]: Removed debugging lines.

**Comment by kunjsong01** [comment]
### Update on support for basic OOP features: 
1. default constructor: (WIP)
  a). ~~zero-arg without side-effect~~ (DONE, Tested in Constructor1-1)
  b). zero-arg with side-effect (WIP)
  c). parameterized (TODO)
  d). parameterized with initializer list (TODO)
2. ~~member access, dot operator~~ (DONE, Tested in Constructor1-8*)
3. member function call (TODO)
  a). setters and getters with side-effect on class member
  b). any public method with side-effect on global variable

### Next step (in another PR): 
Issue #720 [[ESBMC++ Development Tracking] Wrong irept to encode CXXThisExpr](https://github.com/esbmc/esbmc/issues/720)

File: src/clang-cpp-frontend/clang_cpp_convert.h [hunk 12]
```diff
@@ -16,14 +16,46 @@ class clang_cpp_convertert : public clang_c_convertert
   virtual ~clang_cpp_convertert() = default;
 
 protected:
+  typedef std::unordered_map<std::size_t, std::pair<std::string, typet>>
+    this_mapt;
+  this_mapt this_map;
+
   bool get_decl(const clang::Decl &decl, exprt &new_expr) override;
 
+  // Get reference to declared variabales or functions, e.g:
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.h]: Can I ask you to follow the standard for comments as in: https://github.com/esbmc/esbmc/blob/master/src/goto-symex/goto_symex_state.h?

File: src/clang-cpp-frontend/clang_cpp_convert.h [hunk 13]
```diff
@@ -16,14 +16,46 @@ class clang_cpp_convertert : public clang_c_convertert
...
+  bool
+  get_function_body(const clang::FunctionDecl &fd, exprt &new_expr) override;
+
+  // Get function params for C++
+  // contains parsing routines specific to C++ class member functions
+  bool get_function_params(
+    const clang::FunctionDecl &fd,
+    code_typet::argumentst &params) override;
+
+  // Add implicit `this' when parsing C++ class member functions, e.g:
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.h]: Can I ask you to follow the standard for comments as in: https://github.com/esbmc/esbmc/blob/master/src/goto-symex/goto_symex_state.h?

File: src/clang-c-frontend/clang_c_convert.h [hunk 14]
```diff
@@ -85,13 +87,26 @@ class clang_c_convertert
...
 
   virtual bool get_var(const clang::VarDecl &vd, exprt &new_expr);
 
   virtual bool get_function(const clang::FunctionDecl &fd, exprt &new_expr);
 
-  bool get_function_params(const clang::ParmVarDecl &pd, exprt &param);
+  virtual bool
+  get_function_body(const clang::FunctionDecl &fd, exprt &new_expr);
+
+  // Parse function parameters
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.h]: Can I ask you to follow the standard for comments as in: https://github.com/esbmc/esbmc/blob/master/src/goto-symex/goto_symex_state.h?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

**Review by lucasccordeiro — APPROVED** [review]

File: src/clang-c-frontend/clang_c_adjust.h [hunk 15]
```diff
@@ -42,6 +42,7 @@ class clang_c_adjust
   void adjust_symbol(exprt &expr);
   void adjust_comma(exprt &expr);
   void adjust_builtin_va_arg(exprt &expr);
+  virtual void adjust_decl_block(codet &code);
```
Comment (rafaelsamenezes) [inline src/clang-c-frontend/clang_c_adjust.h:48]: Since this is a virtual method, could you add a description for it?

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

File: src/clang-c-frontend/clang_c_language.cpp [hunk 16]
```diff
@@ -277,7 +277,9 @@ bool clang_c_languaget::preprocess(
 
 bool clang_c_languaget::final(contextt &context, const messaget &msg)
 {
+  // populate symbol value using c2goto operational models
```
Comment (fbrausse) [inline src/clang-c-frontend/clang_c_language.cpp]: I don't understand this comment: which symbol value gets populated?

File: src/clang-c-frontend/clang_c_convert.h [hunk 17]
```diff
@@ -64,6 +64,8 @@ class clang_c_convertert
   static void
   gen_typecast_to_union(exprt &dest, const typet &type, const messaget &msg);
 
+  std::string get_decl_name(const clang::NamedDecl &nd);
```
Comment (fbrausse) [inline src/clang-c-frontend/clang_c_convert.h:67]: make it `static` as before?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 18]
```diff
@@ -622,3 +877,67 @@ bool clang_cpp_convertert::get_template_decl(
 
   return false;
 }
+
+bool clang_cpp_convertert::search_this_map(
+  const std::size_t address,
+  this_mapt::iterator &this_it)
```
Comment (fbrausse) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: Is this separate function really necessary? The expression `this_map.find(address) != this_map.end()` is equivalent to it.

**Review by fbrausse — COMMENTED** [review]
Besides minor comments LGTM. Very nice development.

File: src/clang-c-frontend/clang_c_adjust.h (hunk 15 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_adjust.h:48]: Done. Added a comment.

File: src/clang-c-frontend/clang_c_language.cpp (hunk 16 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_language.cpp]: Done. Removed the comment.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 18 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp]: Done. Removed

File: src/clang-c-frontend/clang_c_convert.h (hunk 17 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.h:67]: I didn't manage to to use the `static std::string get_decl_name` function  in `clang_cpp_convertert` class. 

I tried `clang_c_convertert::get_decl_name(fd)`, but got linker error: 
```
/usr/bin/ld: src/clang-cpp-frontend/libclangcppfrontend.a(clang_cpp_convert.cpp.o): in function `clang_cpp_convertert::get_expr(clang::Stmt const&, exprt&)':
../src/clang-cpp-frontend/clang_cpp_convert.cpp:617: undefined reference to `clang_c_convertert::get_decl_name[abi:cxx11](clang::NamedDecl const&)'
/usr/bin/ld: src/clang-cpp-frontend/libclangcppfrontend.a(clang_cpp_convert.cpp.o): in function `clang_cpp_convertert::build_member_from_component(clang::FunctionDecl const&, exprt&)':
../src/clang-cpp-frontend/clang_cpp_convert.cpp:704: undefined reference to `clang_c_convertert::get_decl_name[abi:cxx11](clang::NamedDecl const&)'
```

Hence, changed it to a public function.

Comment (fbrausse) [reply src/clang-c-frontend/clang_c_convert.h:67]: Public is fine; I meant
```diff
diff --git a/src/clang-c-frontend/clang_c_convert.h b/src/clang-c-frontend/clang_c_convert.h
index b8f9560e1e..8c1c43ae51 100644
--- a/src/clang-c-frontend/clang_c_convert.h
+++ b/src/clang-c-frontend/clang_c_convert.h
@@ -64,7 +64,7 @@ public:
   static void
   gen_typecast_to_union(exprt &dest, const typet &type, const messaget &msg);
 
-  std::string get_decl_name(const clang::NamedDecl &nd);
+  static std::string get_decl_name(const clang::NamedDecl &nd);
 
 protected:
   clang::ASTContext *ASTContext;
```

**Comment by mikhailramalho** [comment]
This is not yet ready for merging.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 19]
```diff
@@ -214,6 +214,9 @@ bool clang_c_convertert::get_decl(const clang::Decl &decl, exprt &new_expr)
     const clang::TranslationUnitDecl &tu =
       static_cast<const clang::TranslationUnitDecl &>(decl);
 
+    msg.debug(fmt::format(
+      "TranslationUnit has a total of {} top-level declarations",
+      std::distance(std::cbegin(tu.decls()), std::cend(tu.decls()))));
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_convert.cpp:219]: Remove.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 20]
```diff
@@ -359,6 +362,9 @@ bool clang_c_convertert::get_struct_union_class_fields(
   const clang::RecordDecl &recordd,
   struct_union_typet &type)
 {
+  msg.debug(fmt::format(
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_convert.cpp:365]: Remove.

File: src/clang-c-frontend/clang_c_language.cpp [hunk 21]
```diff
@@ -278,6 +278,7 @@ bool clang_c_languaget::preprocess(
 bool clang_c_languaget::final(contextt &context, const messaget &msg)
 {
   add_cprover_library(context, msg, this);
+  // add __ESBMC__main symbol
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_language.cpp:281]: add ->  adds

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 22]
```diff
@@ -12,12 +12,14 @@
...
 #include <clang/Index/USRGeneration.h>
 #include <clang/Frontend/ASTUnit.h>
+#include <clang/AST/ParentMapContext.h>
 #pragma GCC diagnostic pop
 
 #include <clang-cpp-frontend/clang_cpp_convert.h>
 #include <util/expr_util.h>
 #include <util/std_code.h>
 #include <util/std_expr.h>
+#include <fmt/core.h>
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:22]: Remove.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 23]
```diff
@@ -277,11 +279,49 @@ bool clang_cpp_convertert::get_struct_union_class_methods(
...
     return false;
 
-  for(const auto *decl : cxxrd->methods())
+  // [Q: base class as in C++ inheritance?]
+  for(const auto &decl : cxxrd->bases())
   {
-    exprt dummy;
-    if(get_decl(*decl, dummy))
-      return true;
+    assert(
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:285]: Remove.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 24]
```diff
@@ -277,11 +279,49 @@ bool clang_cpp_convertert::get_struct_union_class_methods(
...
+    if(decl->getKind() == clang::Decl::Field)
+      continue;
+
+    struct_typet::componentt comp;
+
+    if(
+      const clang::FunctionTemplateDecl *ftd =
+        llvm::dyn_cast<clang::FunctionTemplateDecl>(decl))
+    {
+      assert(!"come back and continue - got a tempalte function");
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:303]: Remove.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 25]
```diff
@@ -572,6 +636,197 @@ bool clang_cpp_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
...
+  // object around it
+  if(objectDecl != nullptr)
+  {
+    new_expr.swap(call);
+  }
+  else
+  {
+    assert(
+      !"come back and continue - got new objectDecl, what about side_effect?");
+  }
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:687]: Remove.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 26]
```diff
@@ -572,6 +636,197 @@ bool clang_cpp_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
...
+  std::size_t address = reinterpret_cast<std::size_t>(fd.getFirstDecl());
+
+  this_mapt::iterator it;
+  if(this_map.find(address) == this_map.end())
+  {
+    msg.error(fmt::format(
+      "Pointer `this' for method {} was not added to scope",
+      clang_c_convertert::get_decl_name(fd)));
+    abort();
+  }
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:706]: Remove.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 27]
```diff
@@ -572,6 +636,197 @@ bool clang_cpp_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
...
+  // if it's a constructor, check for initializers
+  if(fd.getKind() == clang::Decl::CXXConstructor)
+  {
+    const clang::CXXConstructorDecl &cxxcd =
+      static_cast<const clang::CXXConstructorDecl &>(fd);
+
+    if(cxxcd.init_begin() != cxxcd.init_end())
+    {
+      assert(!"got ctor initializers");
+      assert(body.has_operands());
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:739]: Remove.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 28]
```diff
@@ -622,3 +877,56 @@ bool clang_cpp_convertert::get_template_decl(
...
+      return true;
+
+    break;
+  }
+
+  default:
+  {
+    msg.error(fmt::format(
+      "Conversion of unsupported clang decl ref: {}", decl.getDeclKindName()));
+    assert(!"add more cases...");
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:921]: Remove.

**Review by mikhailramalho — COMMENTED** [review]
Please, use the message interface to report errors and then abort, instead of using asserts.

**Comment by kunjsong01** [comment]
> Please, use the message interface to report errors and then abort, instead of using asserts.

ACK. I'll replace them with `msg.error`. 

As for the other comments, since Github does allow the user to re-open a closed PR. I'll create a new PR to address them.

**Comment by mikhailramalho** [comment]
Ouch, I just saw that my revert pushed several revert commits to the repo. Sorry about that.

I removed the revert commits from master, so @kunjsong01, you can open a PR to fix only the comments I left.

@lucasccordeiro @rafaelsamenezes where is that document we discussed our development process? Maybe we can add a rule that if a PR is already merged (unless it breaks stuff considerably), it shouldn't be reverted. Fixes should be included in a new PR.

**Comment by fbrausse** [comment]
@mikhailramalho there is #502 and #519, maybe it's one of those.

**Comment by lucasccordeiro** [comment]
@mikhailramalho: In this PR, we have followed the recommendation of the development process. We had three reviewers (Lucas, @fbrausse and @rafaelsamenezes) who provided comments and accepted @kunjian's PR. Before I merged this PR, I spoke with @kunjian to check whether it was ready for merging.

From my perspective, it was a bit unpleasant to revert the entire PR because @kunjsong01 did not address the comments you gave him face-to-face. This action has caused a significant rework for @kunjsong01 since he needs to create another PR from scratch. As far as I understand, your comments about the message interface could easily be implemented in a separate PR.

@fbrausse, @rafaelsamenezes, @kunjsong01: What are your thoughts here?

**Comment by rafaelsamenezes** [comment]
> I removed the revert commits from master, so @kunjsong01, you can open a PR to fix only the comments I left.

They are still there.

> Where is that document we discussed our development process?

It is here: https://github.com/esbmc/esbmc/pull/519

Could I ask whether @mikhailramalho, @fbrausse , @lucasccordeiro, and @kunjsong01 could review again and help me to increment it?

> What are your thoughts here?

1. Remove the revert commits. This way @kunjsong01 does not need to redo this PR
2. Create an issue with the comments.
3. Refrain from reverts and/or rebases over the master branch without opening an issue first.

**Comment by kunjsong01** [comment]
Fixed the outstanding comments in PR https://github.com/esbmc/esbmc/pull/733

**Pull request merged**